### PR TITLE
Balance change notifications

### PR DESCRIPTION
**Pull request by oxarbitrage**
This is a signal/slot approach to do part 2 of #4279 (comment)

First, it moves 2 functions that we are going to need from rpcwallet.cpp to wallet.cpp. This are getBalanceTaddr and getBalanceZaddr.

Then, signal and slots are created to log balance changes in the 3 addresses type(sprout, sapling, transparent).
I left coinbase balance outside but can be added, not sure if it is needed.

A connection to the signal is created once at wallet load(at CWallet::CWallet::InitLoadWallet), then slot is called at every transaction added to the wallet(at CWallet::AddToWallet).

### CODE REVIEW

**Comment by daira** [comment]
Removed this part of the comment that goes into the merge commit:
> Note: I accidentally deleted the branch and the action closed original PR at https://github.com/zcash/zcash/pull/4413 so had to open a new one. I am sorry about that.

(That's completely fine.)

File: src/wallet/wallet.h [hunk 1]
```diff
@@ -1510,5 +1517,67 @@ class AddSpendingKeyToWallet : public boost::static_visitor<SpendingKeyAddResult
...
+            Log(transparent, address, pwalletMain->getBalanceTaddr(address, 0));
+        }
+
+        // sprout
+        std::set<libzcash::SproutPaymentAddress> sproutAdresses;
+        pwalletMain->GetSproutPaymentAddresses(sproutAdresses);
+        for (auto& addr : sproutAdresses) {
+            if (HaveSpendingKeyForPaymentAddress(pwalletMain)(addr)) {
+                const auto& address = EncodePaymentAddress(addr);
+                Log(sprout, address, pwalletMain->getBalanceZaddr(EncodePaymentAddress(addr), 0));
```
[the remark this answers is not in the document]
Comment (oxarbitrage) [reply src/wallet/wallet.h]: D'oh! Thanks, fixed at https://github.com/zcash/zcash/pull/4414/commits/b6bfa0696237cd1e6f679bae030b89608f48b8ae

**Comment by oxarbitrage** [comment]
Thank you for the review @daira , i am also worried about the performance. I just solved the minor changes and now i am after your suggestion to improve performance.

**Comment by oxarbitrage** [comment]
I am trying a new approach at https://github.com/zcash/zcash/pull/4414/commits/6d28b87af689b859b9eb209f9a2925a81014d8d8

Basically i refactored the implementation of `GetFilteredNotes` to use a transaction instead of iterating through all available in the wallet. This `wtx` is optional in higher level overloads calls to `GetFilteredNotes` or `getBalanceZaddr`, when not present current functionality of looping in the wallet is done while when present we only compute inside that transaction.

It is still dirty code but i hope good enough to show the concept. Please let me know what do you think.

**Comment by oxarbitrage** [comment]
Replaced by https://github.com/zcash/zcash/pull/4423

**Pull request closed without merging**